File: openfeature-provider/java/src/main/java/com/spotify/confidence/UnsupportedMaterializationStore.java
package com.spotify.confidence;

import java.util.List;
import java.util.Set;
import java.util.concurrent.CompletionStage;

/**
 * A store that doesn't support the feature and where the provider resorts to use gRPC to resolve
 * flags when the WASM resolver encounters missing materializations. This provides a fallback to the
 * remote Confidence service.
 */
final class UnsupportedMaterializationStore implements MaterializationStore {

  @Override
  public CompletionStage<List<ReadResult>> read(List<? extends ReadOp> ops) {
    throw new MaterializationNotSupportedException();
  }

  @Override
  public CompletionStage<Void> write(Set<? extends WriteOp> ops) {
    throw new MaterializationNotSupportedException();
  }
}

File: openfeature-provider/go/confidence/unsupported_materialization_store.go
package confidence

import (
	"context"
)

// UnsupportedMaterializationStore is a MaterializationStore implementation that always
// returns MaterializationNotSupportedError. This is the default store used by the provider
// to trigger fallback to remote gRPC resolution when materializations are needed.
//
// This allows the Confidence service to manage materializations server-side,
// requiring no additional client-side setup.
type UnsupportedMaterializationStore struct{}

// NewUnsupportedMaterializationStore creates a new UnsupportedMaterializationStore.
func NewUnsupportedMaterializationStore() *UnsupportedMaterializationStore {
	return &UnsupportedMaterializationStore{}
}

// Read always returns MaterializationNotSupportedError to trigger gRPC fallback.
func (u *UnsupportedMaterializationStore) Read(ctx context.Context, ops []ReadOp) ([]ReadResult, error) {
	return nil, &MaterializationNotSupportedError{
		Message: "materialization read not supported, falling back to remote resolution",
	}
}

// Write always returns MaterializationNotSupportedError.
func (u *UnsupportedMaterializationStore) Write(ctx context.Context, ops []WriteOp) error {
	return &MaterializationNotSupportedError{
		Message: "materialization write not supported",
	}
}

// Close is a no-op for UnsupportedMaterializationStore.
func (u *UnsupportedMaterializationStore) Close() error {
	return nil
}
